### PR TITLE
feature (extras/kms): add support to manage the cache of wrappers

### PR DESCRIPTION
**Pull request by jimlambrt**

### CODE REVIEW

**Review by johanbrandhorst — APPROVED** [review]
Nice!

File: extras/kms/kms.go [hunk 1]
```diff
@@ -40,15 +40,20 @@ type Kms struct {
...
 // New takes in a reader, writer and a list of key purposes it will support.
 // Every kms will support a KeyPurposeRootKey by default and it doesn't need to
-// be passed in as one of the supported purposes.  No options are currently
-// supported.
-func New(r dbw.Reader, w dbw.Writer, purposes []KeyPurpose, _ ...Option) (*Kms, error) {
+// be passed in as one of the supported purposes.
+//
+// The WithCache(bool) option is supported and if enabled then a cache of the
+// wrappers is enabled.  Once enabled, the cache will speed up calls to
+// GetWrapper(...) but the caller also becomes responsible to reset/clear the
```
Comment (tmessi) [inline extras/kms/kms.go:54]: how does a scope get deleted from the database?

File: extras/kms/migrations/postgres/04_keys.up.sql [hunk 2]
```diff
@@ -56,7 +56,7 @@ create table kms_data_key (
     references kms_root_key(private_id)
     on delete cascade
     on update cascade,
-  purpose text not null unique
+  purpose text not null
```
Comment (tmessi) [inline extras/kms/migrations/postgres/04_keys.up.sql:59]: Why is this change needed here?

File: extras/kms/kms.go (hunk 1 above)
Comment (jimlambrt) [reply extras/kms/kms.go:54]: that's outside the "scope" of this package.

File: extras/kms/migrations/postgres/04_keys.up.sql (hunk 2 above)
Comment (jimlambrt) [reply extras/kms/migrations/postgres/04_keys.up.sql:59]: in the course of testing the new cache bits I attempted to add a `database` DEK to two scopes and it failed which is a defect.

Comment (tmessi) [reply extras/kms/migrations/postgres/04_keys.up.sql:59]: Ah, I see, a few lines below we have `unique (root_key_id, purpose) -- there can only be one dek for a specific purpose per root key` which is the correct constraint here.

**Review by tmessi — APPROVED** [review]
